File: src/main/java/com/tms/csp/transforms/FlattenTopLevelAnds.java
package com.tms.csp.transforms;

import com.tms.csp.argBuilder.ArgBuilder;
import com.tms.csp.ast.Exp;
import com.tms.csp.ast.Op;

import java.util.List;

public class FlattenTopLevelAnds extends BaseTransformer {

    @Override
    protected Exp local(Exp in) {
        return flattenTopLevelAnds(in);
    }

    @Override
    protected boolean executeLocal(Exp in) {
        if (in.isAnd()) {
            for (Exp arg : in.argIt()) {
                if (arg.isAnd()) {
                    return true;
                }
            }
        }
        return false;
    }

    private Exp flattenTopLevelAnds(Exp in) {
        boolean anyChange = false;

        ArgBuilder flatter = new ArgBuilder(in.getSpace(), Op.And);
        for (Exp arg : in.args()) {
            if (arg.isAnd()) {
                List<Exp> args = arg.args();
                flatter.addExpIt(args);
                anyChange = true;
            } else {
                flatter.addExp(arg);
            }
        }


        if (anyChange) {
            return flatter.mk();
//            return space.mkAnd(flatter);
        } else {
            return in;
        }

    }

}

File: src/main/java/com/tms/csp/transforms/FlattenTopLevelAnds.kt
package com.tms.csp.transforms

import com.tms.csp.argBuilder.ArgBuilder
import com.tms.csp.ast.Exp
import com.tms.csp.ast.Op

class FlattenTopLevelAnds : BaseTransformer() {

    override fun local(inExp: Exp): Exp? {
        return flattenTopLevelAnds(inExp)
    }

    override fun executeLocal(inExp: Exp): Boolean {
        if (inExp.isAnd) {
            for (arg in inExp.argIt) {
                if (arg.isAnd) {
                    return true
                }
            }
        }
        return false
    }

    private fun flattenTopLevelAnds(inExp: Exp): Exp {
        var anyChange = false

        val flatter = ArgBuilder(inExp.space, Op.And)
        for (arg in inExp.args) {
            if (arg.isAnd) {
                val args = arg.args
                flatter.addExpIt(args)
                anyChange = true
            } else {
                flatter.addExp(arg)
            }
        }


        return if (anyChange) {
            flatter.mk()
            //            return space.mkAnd(flatter);
        } else {
            inExp
        }

    }

}
